### ISSUE TITLE
VGGT W/ PROPE Not Working

### ISSUE THREAD

**Issue by StephenYangjz** [liruilong940607/prope#11]
Hi Ruilong, I have integrated PROPE into (Stream)VGGT, and removed the camera tokens and camera predict head. However, with exactly the same model architecture, I found that the model performs better without PROPE. It's a smaller model w/ dinov3 trained from scratch distilling VGGT in both cases, so I don't think the initilizatino of the model makes the difference. I also trained  it over 64 A100s across pretty massive datasets, so I am wondering if you have any intuitions on why it may not work.

I have also attached 2 visualizations, one with PROPE and the other without PROPE. I believe I have done the right implementatino in terms of removing register tokens and then adding them back; prope the KV cache during inference time, etc. And I assume that PROPE should work naively with a casual mask as well.

Any pointers would be greatly appreciated!

<img width="667" height="147" alt="Image" src="https://github.com/user-attachments/assets/417f570a-5b14-4cf5-83a1-4514cebb3571" />

<img width="640" height="131" alt="Image" src="https://github.com/user-attachments/assets/a1cd32df-28ef-4273-b277-7825331bd2cc" />

**Comment by StephenYangjz** [comment]
@liruilong940607 Thank you so much for getting back to me! I removed the camera tokens and the camera head, and is trying to train a smaller, streamed, and distilled VGGT for 3D reconstruction conditioned on images and camera poses. That's why I think PROPE is helpful, and has been adding it into the global attention layers. I have attached some relevant files that I modified [here](https://drive.google.com/file/d/14fdrRyjBvcZIUTdk65ukbEWx_DBQW_1Q/view?usp=sharing) for your reference. Appreciate it again!

**Comment by liruilong940607** [comment]
A quick skim through your code. This (and similar for q and v) feels weird:

```python
                    k_cur = torch.cat([
                        apply_fn_kv(k_cur[:, :, -image_len:, :]),
                        k_cur[:, :, :prefix_len, :]
                    ], dim=2)

```

Why swapping the order between `prefix` and `cur`? Shouldn't it be

```python
                    k_cur = torch.cat([
                        k_cur[:, :, :prefix_len, :]
                        apply_fn_kv(k_cur[:, :, -image_len:, :]),
                    ], dim=2)
```

**Comment by HaojiLan** [comment]
Hi! @liruilong940607 , appreciate your great work!  When adding camera pose into Alternating-Attention of VGGT,  **implement PROPE in global attention block or frame attention block,** which one is more reasonable? Could you give me some suggestions?

**Comment by StephenYangjz** [comment]
Thanks @liruilong940607. I did some ablations across different datasets and it does seem like VKITTI doesnt have this issue. I investigated into the `viewmat` translations and observed that the the issue indeed comes from Waymo dataset's absolute translations that can go to a couple thousands.

As for the best way to normalize them-- do you think we can simply normalize against each scenario/drive, as we are not relying the network to reason about any absolute map scale? But at the same time, this would mean we are giving up metric information (if there's any flowing through PROPE)... Do you think it's possible to keep any metric info for VGGT to predict pointmaps in a close-to-metric scale? i.e. Does subtracting the center of the drive help?

**Comment by liruilong940607** [comment]
@StephenYangjz Oh that is actually NOT fine if you mean something like `q = rope(prope(q, camera), scale)` or `q = prope(rope(q, scale), camera)`.

Both PROPE and ROPE relies on the dot product between rotated Q and rotate K to achieve **relative** encoding. So if you overlay another encoding on top of it, it will break its property of achieving relative encoding.

However, it might "just works" magically -- who knows -- just give you a heads-up that it will break the underlying math of ROPE / PROPE

**Comment by liruilong940607** [comment]
closing at the question seems resolved. feel free to reopen if there is following up question.
